File: ZKWeb.Toolkits/ZKWeb.Toolkits.ProjectCreator/Model/CreateProjectParameters.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ZKWeb.Toolkits.ProjectCreator.Properties;

namespace ZKWeb.Toolkits.ProjectCreator.Model {
	/// <summary>
	/// Create project parameters
	/// </summary>
	public class CreateProjectParameters {
		/// <summary>
		/// Project template directory
		/// Automatic detect if empty
		/// </summary>
		/// <example>D:\\Project\\ZKWeb\\Tools\\Templates</example>
		public string TemplatesDirectory { get; set; }
		/// <summary>
		/// Project type
		/// </summary>
		/// <example>AspNet</example>
		public string ProjectType { get; set; }
		/// <summary>
		/// Project name
		/// </summary>
		/// <example>ZKWeb.Demo</example>
		public string ProjectName { get; set; }
		/// <summary>
		/// Project description
		/// </summary>
		/// <example>Some Description</example>
		public string ProjectDescription { get; set; }
		/// <summary>
		/// ORM
		/// </summary>
		/// <example>NHibernate</example>
		public string ORM { get; set; }
		/// <summary>
		/// Database
		/// </summary>
		/// <example>PostgreSQL</example>
		public string Database { get; set; }
		/// <summary>
		/// Connection string
		/// </summary>
		/// <example>Server=127.0.0.1;Port=5432;Database=zkweb;User Id=postgres;Password=123456;</example>
		public string ConnectionString { get; set; }
		/// <summary>
		/// If use default plugins, it should be the path of `plugin.collection.json`
		/// otherwise it should be null
		/// </summary>
		/// <example>D:\\Projects\\ZKWeb.Plugins\\plugin.collection.json</example>
		public string UseDefaultPlugins { get; set; }
		/// <summary>
		/// Project output directory
		/// </summary>
		public string OutputDirectory { get; set; }
		/// <summary>
		/// Available product types
		/// </summary>
		public readonly static string[] AvailableProductTypes = new[] {
			"AspNetCore", "AspNet", "Owin"
		};
		/// <summary>
		/// Available ORM
		/// InMemory should only use for test, so it's not here
		/// </summary>
		public readonly static string[] AvailableORM = new[] {
			"Dapper", "EFCore", "MongoDB", "NHibernate"
		};
		/// <summary>
		/// Available databases for specified ORM
		/// </summary>
		public readonly static IDictionary<string, string[]> AvailableDatabases =
			new Dictionary<string, string[]> {
				{ "Dapper", new[] { "MSSQL", "SQLite", "MySQL", "PostgreSQL" } },
				{ "EFCore", new[] { "MSSQL", "SQLite", "MySQL", "PostgreSQL", "InMemory" } },
				{ "MongoDB", new [] { "MongoDB" } },
				{ "NHibernate", new[] { "PostgreSQL", "SQLite", "MySQL", "MSSQL" } }
			};

		/// <summary>
		/// Check parameters
		/// </summary>
		public void Check() {
			if (!AvailableProductTypes.Contains(ProjectType)) {
				throw new ArgumentException(
					Resources.ProjectTypeMustBeOneOf +
					string.Join(", ", AvailableProductTypes));
			} else if (string.IsNullOrEmpty(ProjectName)) {
				throw new ArgumentException(Resources.ProjectNameCantBeEmpty);
			} else if (!AvailableORM.Contains(ORM)) {
				throw new ArgumentException(
					Resources.ORMMustBeOneOf +
					string.Join(", ", AvailableORM));
			} else if (!AvailableDatabases[ORM].Contains(Database)) {
				throw new ArgumentException(
					Resources.DatabaseMustBeOneOf +
					string.Join(", ", AvailableDatabases[ORM]));
			} else if (string.IsNullOrEmpty(ConnectionString)) {
				throw new ArgumentException(Resources.ConnectionStringCantBeEmpty);
			} else if (string.IsNullOrEmpty(OutputDirectory)) {
				throw new ArgumentException(Resources.OutputDirectoryCantBeEmpty);
			}
			if (!string.IsNullOrEmpty(UseDefaultPlugins)) {
				var pluginCollection = PluginCollection.FromFile(UseDefaultPlugins);
				if (!pluginCollection.SupportedORM.Any(o =>
					string.Equals(o, ORM, StringComparison.OrdinalIgnoreCase))) {
					throw new ArgumentException(
						Resources.ThisPluginCollectionOnlySupportTheseORM +
						string.Join(", ", pluginCollection.SupportedORM));
				}
			}
		}
	}
}

File: ZKWeb.Toolkits/ZKWeb.Toolkits.ProjectCreator.Crossplatform.Gui/src/app/CreateProjectParameters.ts
const fs = require("fs");
export class CreateProjectParameters {

    public TemplatesDirectory: string;
    public ProjectType: string;
    public ProjectName: string;
    public ProjectDescription: string;
    public ORM: string;
    public Database: string;
    public ConnectionString: string;
    public UseDefaultPlugins: string;
    public OutputDirectory: string;
    public AvailableProductTypes: string[] = ["AspNetCore", "AspNet", "Owin"];
    public AvailableORM: string[] = ["Dapper", "EFCore", "MongoDB", "NHibernate"];
    public AvailableDatabases: any = {
        Dapper: ["MSSQL", "SQLite", "MySQL", "PostgreSQL"],
        EFCore: ["MSSQL", "SQLite", "MySQL", "PostgreSQL", "InMemory"],
        MongoDB: ["MongoDB"],
        NHibernate: ["PostgreSQL", "SQLite", "MySQL", "MSSQL"],
    };

    public Check(): any {
        if (-1 === this.AvailableProductTypes.indexOf(this.ProjectType)) {
            return {
                isSuccess: false,
                msgPrefix: "ProjectTypeMustBeOneOf",
                args: this.AvailableProductTypes.join(","),
            };
        } else if (!this.ProjectName) {
            return {
                isSuccess: false,
                msgPrefix: "ProjectNameCantBeEmpty",
            };
        } else if (-1 === this.AvailableORM.indexOf(this.ORM)) {
            return {
                isSuccess: false,
                msgPrefix: "ORMMustBeOneOf",
                args: this.AvailableORM.join(","),
            };
        } else if (-1 === this.AvailableDatabases[this.ORM].indexOf(this.Database)) {
            return {
                isSuccess: false,
                msgPrefix: "DatabaseMustBeOneOf",
                args: this.AvailableDatabases[this.ORM].join(","),
            };
        } else if (this.Database !== "InMemory" && !this.ConnectionString) {
            return {
                isSuccess: false,
                msgPrefix: "ConnectionStringCantBeEmpty",
            };
        } else if (!this.OutputDirectory) {
            return {
                isSuccess: false,
                msgPrefix: "OutputDirectoryCantBeEmpty",
            };
        }
        if (this.UseDefaultPlugins) {
            if(!fs.existsSync(this.UseDefaultPlugins)){
                return {
                    isSuccess: false,
                    msgPrefix: "DefaultPluginsFileNotFound",
                };
            }

            const json = fs.readFileSync(this.UseDefaultPlugins, "utf8");
            const pluginCollection =  JSON.parse(json);
            let isOrmExit = false;

            for (const orm of pluginCollection.SupportedORM) {
                if (orm === this.ORM) {
                    isOrmExit = true;
                }
            }

            if (!isOrmExit) {
                return {
                    isSuccess: false,
                    msgPrefix: "ORMMustBeOneOf",
                    args: pluginCollection.SupportedORM.join(","),
                };
            }
        }
        return {
            isSuccess: true,
        };
    }
}
